cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java

```diff
@@ -160,7 +164,8 @@ protected AbstractContext(Program program, ProgramOptions programOptions, CConfi
                             @Nullable MetricsCollectionService metricsService, Map<String, String> metricsTags,
                             SecureStore secureStore, SecureStoreManager secureStoreManager,
                             MessagingService messagingService,
-                            @Nullable PluginInstantiator pluginInstantiator) {
+                            @Nullable PluginInstantiator pluginInstantiator,
+                            boolean shouldSetContextClassloader) {
```

Comment:
I think it's better to just have the `setContextCombinedClassLoader` method be `protected` and override it in the `BasicMapReduceContext`, `BasicMapReduceTaskContext` and `SparkRuntimeContext` to return the preconstructed one.

Reply:
alternatively, have the "user" classloader passed in from the constructor, as it won't change.

Reply:
I can do that but it will likely require a setter, since the mapreduce classloader requires the job's configuration.

Reply:
The `BasicMapReduceContext` already has the `Job` object. See https://github.com/caskdata/cdap/blob/release/4.3/cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java#L222

Reply:
yeah but the conf actually used for the classloader is different. For spark it's strange as well, since the SparkClassLoader takes a context in its constructor.

Reply:
For the Spark case, you can get it by `SparkClassLoader.findFromContext`. And for both MR and Spark, we should wrap it with the weak reference classloader.